src/raftstore/coprocessor/split_check/table.rs

```diff
@@ -0,0 +1,488 @@
+        let mut table_status = TableStatus::default();
+        let skip = before_check(&mut table_status, engine, ctx.region());
+        if !skip {
+            status.table = Some(table_status);
+        }
+    }
+
+    fn on_split_check(
+        &self,
+        _: &mut ObserverContext,
+        status: &mut Status,
+        key: &[u8],
+        _: u64,
+    ) -> Option<Vec<u8>> {
+        if let Some(ref mut table_status) = status.table {
+            check_key(table_status, key)
+        } else {
+            None
+        }
+    }
+}
+
+/// Do some quick checks, true for skipping `check_key`.
+fn before_check(status: &mut TableStatus, engine: &DB, region: &Region) -> bool {
+    if let Ok(true) = is_same_table(region.get_start_key(), region.get_end_key()) {
+        // Region is inside a table, skip for saving IO.
+        return true;
+    }
+
+    let (start_key, end_key) = match bound_keys(engine, region) {
+        Ok(Some((start_key, end_key))) => (start_key, end_key),
+        Ok(None) => return true,
+        Err(err) => {
+            error!(
+                "[region {}] failed to get region bound: {}",
+                region.get_id(),
+                err
+            );
+            return true;
+        }
+    };
+
+    if !keys::validate_data_key(&start_key) || !keys::validate_data_key(&end_key) {
+        return true;
+    }
+
+    let encoded_start_key = keys::origin_key(&start_key);
+    let encoded_end_key = keys::origin_key(&end_key);
+
+    if encoded_start_key.len() < table_codec::TABLE_PREFIX_KEY_LEN ||
+        encoded_end_key.len() < table_codec::TABLE_PREFIX_KEY_LEN
+    {
+        // For now, let us scan region if encoded_start_key or encoded_end_key
+        // is less than TABLE_PREFIX_KEY_LEN.
+        return false;
+    }
+
+    // Table data starts with `TABLE_PREFIX`.
+    // Find out the actual range of this region by comparing with `TABLE_PREFIX`.
+    match (
+        encoded_start_key[..table_codec::TABLE_PREFIX_LEN].cmp(table_codec::TABLE_PREFIX),
+        encoded_end_key[..table_codec::TABLE_PREFIX_LEN].cmp(table_codec::TABLE_PREFIX),
+    ) {
+        // The range does not cover table data.
+        (Ordering::Less, Ordering::Less) | (Ordering::Greater, Ordering::Greater) => true,
+
+        // Following arms matches when the region contains table data.
+        // Covers all table data.
+        (Ordering::Less, Ordering::Greater) => false,
+        // The later part contains table data.
+        (Ordering::Less, Ordering::Equal) => {
+            // It starts from non-table area to table area,
+            // try to extract a split key from `encoded_end_key`, and save it in status.
+            status.last_encoded_table_prefix = to_encoded_table_prefix(encoded_end_key);
+            false
+        }
+        // Region is in table area.
+        (Ordering::Equal, Ordering::Equal) => {
+            if let Ok(true) = is_same_table(encoded_start_key, encoded_end_key) {
```

Comment:
It has been checked at L66.

Reply:
It is different from L66, what L66 checks is region's start_key and end_key, but here are the actually start key and end key that are stored in rocksdb.